lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,278 @@
+    # Groups of subnets are cached under the External IPAM Group name. "IPAM Group" in the snippet below is
+    # the 'section' in phpIPAM. For subnets that have an External IPAM group specified(e.g. "IPAM Group"),
+    # the IP's/mac are cached under the "IPAM Group" key. If External IPAM group is not defined, then they
+    # are cached under the "" key.
+    #
+    # The IP address is only actually reserved in the External IPAM database upon successful host
+    # and/or interface creation.
+    #
+    # In-memory IP cache structure(IPv4 and IPv6):
+    # ===============================
+    # {
+    #   "": {
+    #     "100.55.55.0/24":{
+    #       "00:0a:95:9d:68:10": {"ip": "100.55.55.1", "timestamp": "2019-09-17 12:03:43 -D400"}
+    #     },
+    #   },
+    #   "IPAM Group": {
+    #     "123.11.33.0/24":{
+    #       "00:0a:95:9d:68:33": {"ip": "123.11.33.1", "timestamp": "2019-09-17 12:04:43 -0400"},
+    #       "00:0a:95:9d:68:34": {"ip": "123.11.33.2", "timestamp": "2019-09-17 12:05:48 -0400"},
+    #       "00:0a:95:9d:68:35": {"ip": "123.11.33.3", "timestamp:: "2019-09-17 12:06:50 -0400"}
+    #     }
+    #   },
+    #   "Another IPAM Group": {
+    #     "185.45.39.0/24":{
+    #       "00:0a:95:9d:68:55": {"ip": "185.45.39.1", "timestamp": "2019-09-17 12:04:43 -0400"},
+    #       "00:0a:95:9d:68:56": {"ip": "185.45.39.2", "timestamp": "2019-09-17 12:05:48 -0400"}
+    #     }
+    #   }
+    # }
+    #
+    # Params: 1. subnet:  The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24", IPv6 - "2001:db8:abcd:12::/124")
+    #         2. mac:     The mac address of the interface obtaining the IP address for
+    #         3. group:   The group in External IPAM that the subnet belongs to(e.g. This would be "Section" in phpIPAM)
+    #
+    # Returns: A hash with the next available IP in the "data" field, or a hash with "error" key
+    #          containing error message, and an HTTP code
+    #
+    # Examples:
+    #   Responses if success:
+    #     IPv4: {"code": 200, "data": "100.55.55.3"}
+    #     IPv6: {"code": 200, "data": "2001:db8:abcd:12::1"}
+    #   Response if missing required params:
+    #     {"code": 400, "error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)"]}
+    #   Response if subnet does not exist:
+    #     {"code": 404, "error": "The specified subnet does not exist in External IPAM."}
+    #   Response if there are no free addresses:
+    #     {"code": 404, "error": "No free addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def next_ip(subnet, mac, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse get("/subnet/#{subnet}/next_ip?mac=#{mac}&group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
```

Comment:
Here, I'd probably simply raise RuntimeError, it will be wrapped in ProxyException few lines below. Also do not use i18n on the output coming from a service - there will be no translation source for these kind of messages. So simply:

    raise(response['error']) if response['error'].present?